src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,354 @@
+            lock.unlock();
+            finish();
+            return;
+        };
+
+        m_fbo = std::make_unique<QOpenGLFramebufferObject>(
+                m_screenInfo.size, QOpenGLFramebufferObject::CombinedDepthStencil);
+
+        GLenum glError;
+        glError = m_context->functions()->glGetError();
+
+        VERIFY_OR_DEBUG_ASSERT(glError == GL_NO_ERROR) {
+            kLogger.warning() << "GLError: " << glError;
+            lock.unlock();
+            finish();
+            return;
+        };
+
+        VERIFY_OR_DEBUG_ASSERT(m_fbo->isValid()) {
+            kLogger.warning() << "Failed to initialize FBO";
+            lock.unlock();
+            finish();
+            return;
+        };
+
+        m_quickWindow->setGraphicsDevice(QQuickGraphicsDevice::fromOpenGLContext(m_context.get()));
+
+        VERIFY_OR_DEBUG_ASSERT(m_renderControl->initialize()) {
+            kLogger.warning() << "Failed to initialize redirected Qt Quick rendering";
+            lock.unlock();
+            finish();
+            return;
+        };
+
+        m_quickWindow->setRenderTarget(QQuickRenderTarget::fromOpenGLTexture(m_fbo->texture(),
+                m_screenInfo.size));
+
+        m_quickWindow->setGeometry(0, 0, m_screenInfo.size.width(), m_screenInfo.size.height());
+    }
+
+    m_nextFrameStart = mixxx::Time::elapsed();
+
+    m_renderControl->polishItems();
+    m_renderControl->beginFrame();
+
+    if (m_pControllerEngine) {
+        m_pControllerEngine->pause();
+    }
+
+    VERIFY_OR_DEBUG_ASSERT(m_renderControl->sync()) {
+        kLogger.warning() << "Couldn't sync the render control.";
+        lock.unlock();
+        finish();
+        if (m_pControllerEngine) {
+            m_pControllerEngine->resume();
+        }
+
+        return;
+    };
+
+    if (m_pControllerEngine) {
+        m_pControllerEngine->resume();
+    }
+    QImage fboImage(m_screenInfo.size, m_screenInfo.pixelFormat);
+
+    VERIFY_OR_DEBUG_ASSERT(m_fbo->bind()) {
+        kLogger.warning() << "Couldn't bind the FBO.";
+    }
+    GLenum glError;
+    m_context->functions()->glFlush();
+    glError = m_context->functions()->glGetError();
+    VERIFY_OR_DEBUG_ASSERT(glError == GL_NO_ERROR) {
+        kLogger.warning() << "GLError: " << glError;
+        lock.unlock();
+        finish();
+        return;
+    }
+    if (m_screenInfo.endian != std::endian::native) {
+        m_context->functions()->glPixelStorei(GL_PACK_SWAP_BYTES, GL_TRUE);
+    }
+    glError = m_context->functions()->glGetError();
+    VERIFY_OR_DEBUG_ASSERT(glError == GL_NO_ERROR) {
+        kLogger.warning() << "GLError: " << glError;
+        lock.unlock();
+        finish();
+        return;
+    }
+
+    QDateTime timestamp = QDateTime::currentDateTime();
+    m_renderControl->render();
+    m_renderControl->endFrame();
+
+    // Flush any remaining GL errors
+    while (m_context->functions()->glGetError()) {
+    }
+    m_context->functions()->glReadPixels(0,
```

Comment:
Ah! I had `PerformanceTimer` (or `ScopedTimer`, can't remember) till I cleaned up the PR a few month ago. Do you think it is valuable to get it back and keep it in the long run?

Reply:
I think so yeah. It makes sense for mapping authors to easily optimize their "transformFunction" and scene complexity.

Reply:
The time it takes to put the frame into memory is more interesting to us IMO because I remember that part being mentioned to be slow in the Qt docs. at least `QQuickItem::grabToImage` is.

Reply:
I don't use `QQuickItem::grabToImage` directly for this reason, and also because we have a very limited flexibility when it comes to pixel format.

Reply:
Yeah definitely makes sense. But there must be some reason why grabToImage is slow, I'm wondering if this also applies in this solution.